File: m03_robot_description/m03_tf2_with_py/m03_tf2_with_py/turtle_listener.py
# !/usr/bin/env python3

# ----------------------- Import of standard libraries ------------------------
import math
import random
import time

# -------------------- Import ROS2 rclpy immplementations ---------------------
import rclpy
from rclpy.node import Node

# -------------------- Import tf2 related packages info -----------------------
from tf2_ros import TransformException
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener

# ---------------- Import of ROS2 msg and srv from other packages -------------
from geometry_msgs.msg import Twist
from turtlesim.srv import Spawn


# ----------------- Listener class of transforms (TF2)-------------------------
class FrameListener(Node):
    """
    Class oriented to listening to a given transform and the movements that are
    applied to it.

    Attributes
    ---
    target_frame : String
        Indicates the frame to listen (frame with the tf of interest)
        
    tf_buffer : Transform Buffer
        Buffer that will storage the transform
        
    tf_listener : Transform Listener
        Listener (or subscriber) that will receive the tf info of interest.
        
    spawner : ROS2 Client
        Client that will call the /spawn service to create a new turtle
    
    publisher : ROS2 Publisher
        Publisher that will send data through /cmd_vel of the target turtle.
        
    timer : ROS2 timer
        Timer that will link a callback to listen to the transform
    
    Methods
    ---
    on_timer()
        Callback for the timer
    """
    def __init__(self):
        """
        Constructor that initialize the node with the name 
        'turtle_tf2_frame_listener', create a transform listener and
        instance a cmd_vel publisher for turtles.
        """
        # Initialize node
        super().__init__('turtle_tf2_frame_listener')

        # Declare and acquire `target_frame` parameter
        self.target_frame = self.declare_parameter(
          'target_frame', 'turtle1').get_parameter_value().string_value

        # Create a buffer and a transform listener
        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        # Create a client to spawn a turtle and the flags to check the spawn.
        self.spawner = self.create_client(Spawn, 'spawn')
        self.turtle_spawning_service_ready = False
        self.turtle_spawned = False

        # Create turtle2 velocity publisher
        self.publisher = self.create_publisher(Twist, 'turtle2/cmd_vel', 1)

        # Call on_timer function every second
        self.timer = self.create_timer(1.0, self.on_timer)

    def on_timer(self):
        # Store frame names, then they will be computed
        from_frame_rel = self.target_frame
        to_frame_rel = 'turtle2'

        if self.turtle_spawning_service_ready:
            if self.turtle_spawned:
                # Look up for the transform between the two frames and get the info
                try:
                    t = self.tf_buffer.lookup_transform(
                        to_frame_rel,
                        from_frame_rel,
                        rclpy.time.Time())
                except TransformException as ex:
                    self.get_logger().info(
                        f'Could not transform {to_frame_rel} to {from_frame_rel}: {ex}')
                    return

                # Create Twist message with the stimation to reach the other turtle
                msg = Twist()
                scale_rotation_rate = 1.0
                msg.angular.z = scale_rotation_rate * math.atan2(
                    t.transform.translation.y,
                    t.transform.translation.x)
                scale_forward_speed = 0.5
                msg.linear.x = scale_forward_speed * math.sqrt(
                    t.transform.translation.x ** 2 +
                    t.transform.translation.y ** 2)
                
                # Publish velocity command and start following turtle
                self.publisher.publish(msg)
                self.get_logger().info("Following turtle, you can check tfs")
                
            else:
                # Additional check on spawn call of the turtle
                if self.result.done():
                    self.get_logger().info(
                        f'Successfully spawned {self.result.result().name}')
                    self.turtle_spawned = True
                else:
                    self.get_logger().info('Spawn was not finished')
        else:
            # In case the turtle doesn't exist, it spawns one in a random pos
            if self.spawner.service_is_ready():
                # Set up random position and fill out the request
                random.seed(int(time.time()))
                request = Spawn.Request()
                request.name = 'turtle2'
                request.x = float(random.randrange(0, 10))
                request.y = float(random.randrange(0, 10))
                request.theta = float(random.uniform(0, math.pi))
                
                # Made the request and update flag
                self.result = self.spawner.call_async(request)
                self.turtle_spawning_service_ready = True
                self.get_logger().info("Spawned new listener turtle")
            else:
                # Check if the service is ready
                self.get_logger().info('Turtlesim srvs seems to be offline')


def main():
    # Initialize ROS Client Library for Python
    rclpy.init()
    
    # Create a logger for displaying info
    logger = rclpy.logging.get_logger('logger')
    logger.info("Initialized: Listener of turtles tfs.")
    
    # Instance node
    node = FrameListener()
    
    # Spin node until a keyboard interrupt is detected
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    
    # Close program 
    logger.info("Closing: Listener of turtles tfs.")
    rclpy.shutdown()

File: m03_robot_description/m03_tf2_with_cpp/src/turtle_listener.cpp
// ------------------------------ REQUIRED STANDARD HEADERS --------------------
#include <chrono>     // Time management with different precisions
#include <functional> // For functions and hash related
#include <memory>     // Dynamic memory management
#include <string>     // String utilities

// ------------------------------ ROS2 RELATED HEADERS ------------------------
#include "rclcpp/rclcpp.hpp"             // ROS2 Client Library for C++
#include "tf2/exceptions.h"              // Transform exceptions
#include "tf2_ros/transform_listener.h"  // Transform listener
#include "tf2_ros/buffer.h"              // Transform buffer

// ----------------------------- ROS2 MSGS DEPENDENCIES -----------------------
#include "geometry_msgs/msg/transform_stamped.hpp"   // Tf with time stamped
#include "geometry_msgs/msg/twist.hpp"               // Velocity command

// ----------------------------- ROS2 SRVS DEPENDENCIES -----------------------
#include "turtlesim/srv/spawn.hpp"   // Service to create new turtles

// ----------------------------- NAMESPACES CONSIDERATIONS --------------------
using namespace std::chrono_literals;  // For time user defined literals

// ----------------------- TF LISTERNER CLASS IMPLEMENTATION -----------------

/**
 * Creates nodes that will listen to the tranforms in order to implement a tf
 * follower with turtles.
 */
class TurtleListener : public rclcpp::Node
{
public:
    /**
     * User defined constructor that initialize the node with the name 
     * turtle_tf2_frame_lister, declares a target frame, a transform buffer, 
     * a turtle spawner server client and a cmd_vel publisher to create the
     * transform listerner linked to a timer callback.
     */
    TurtleListener()
        : Node("turtle_tf2_frame_listener"),
          turtle_srv_is_ready_(false),
          turtle_spawned_(false)
    {
        // Consider declared parameter for the name
        target_frame_ = this->declare_parameter<std::string>("target_frame",
            "turtle1");
        
        // Buffer that has current time information
        tf_buf_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());

        // Instance transform listener that considers a pointer to the buffer
        tf_listen_ = std::make_shared<tf2_ros::TransformListener>(*tf_buf_);

        // Instance a spawner server client
        spawner_ = this->create_client<turtlesim::srv::Spawn>("spawn");

        // Instance a cmd_vel publisher to create follower of tfs
        cmd_vel_pub_ = this->create_publisher<geometry_msgs::msg::Twist>(
            "follower/cmd_vel", 1);

        // Implement timer with it sproper callback
        timer_ = this->create_wall_timer(1s, 
            std::bind(&TurtleListener::on_timer, this));
    }
private:
    /**
     * Callback for the timer that will lookup for the transform between the 
     * turtles, then interpret the distnace and make the other turtle follow 
     * the first one.
     * 
     * If the turtle that follows the first one doesn't extis, it will spawn a
     * new one.
     * 
     */
    void on_timer()
    {
        // Obtain and assign names of the transform of interest
        std::string fromFrameRel = target_frame_.c_str();
        std::string toFrameRel = "follower";

        // Check if the spawn service is ready
        if(turtle_srv_is_ready_)
        {
            // Check if the follower turtle spawned
            if(turtle_spawned_)
            {
                // Declare stamped transform
                geometry_msgs::msg::TransformStamped t_stamp;

                try
                {
                    // Look up for transform as soon as it is available
                    t_stamp = tf_buf_->lookupTransform(
                        toFrameRel, fromFrameRel, tf2::TimePointZero);
                }
                catch(const tf2::TransformException & ex)
                {
                    // Display message in case of error for tfs loop up
                    RCLCPP_INFO(
                        this->get_logger(), "Could find transform between"
                        "%s to %s: %s", toFrameRel.c_str() , 
                        fromFrameRel.c_str(), ex.what()
                    );
                    return;
                }

                // Declare twist message for velocity commands
                geometry_msgs::msg::Twist msg;

                // Assign rotation based on the x and y distance between turtle
                static const double scaleRotationRate = 1.0;
                msg.angular.z = scaleRotationRate * atan2(
                    t_stamp.transform.translation.y,
                    t_stamp.transform.translation.x);

                // Assign linear velocity based on Pitagoras Theorem scaled
                static const double scaleForwardSpeed = 0.5;
                msg.linear.x = scaleForwardSpeed * sqrt(
                    pow(t_stamp.transform.translation.x, 2) +
                    pow(t_stamp.transform.translation.y, 2));

                // Publish velocity command
                cmd_vel_pub_->publish(msg);
            }
            else
            {
                // Log to show that turtle was spawned
                RCLCPP_INFO(this->get_logger(), "Succesfully spawned turtle");
                turtle_spawned_ = true;
            }
        }
        else
        {
            // If the service is ready....
            if (spawner_->service_is_ready())
            {
                // Make the request to spawn a turtle in the givne position
                auto request = 
                    std::make_shared<turtlesim::srv::Spawn::Request>();
                request->x = 4.0;
                request->y = 2.0;
                request->theta = 0.0;
                request->name = "follower";

                // Create future response
                using ServiceResponseFuture = 
                    rclcpp::Client<turtlesim::srv::Spawn>::SharedFuture;
                
                // Lambda for checking that the turtle was created
                auto resp_rec_callback = [this](ServiceResponseFuture future)
                {
                    auto result = future.get();
                    if(strcmp(result->name.c_str(), "follower") == 0)
                    {
                        turtle_srv_is_ready_ = true;
                    }
                    else
                    {
                        RCLCPP_ERROR(this->get_logger(), "Spawn didn't work");
                    }
                };

                // Send request to spawn turtle with verification
                auto result = spawner_->async_send_request(request, 
                    resp_rec_callback);
            }
            else
            {
                RCLCPP_INFO(this->get_logger(), "Service is not ready");
            }
        }
    }
    // Declare flags of services
    bool turtle_srv_is_ready_;
    bool turtle_spawned_;

    // Declare string for parameter
    std::string target_frame_;

    // Declare ros related private attributes for service, timer, publisher, tf
    rclcpp::Client<turtlesim::srv::Spawn>::SharedPtr spawner_{nullptr};
    rclcpp::TimerBase::SharedPtr timer_{nullptr};
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_pub_{nullptr};
    std::shared_ptr<tf2_ros::TransformListener> tf_listen_{nullptr};
    std::unique_ptr<tf2_ros::Buffer> tf_buf_;
};

// ------------------------- MAIN IMPLEMENTATION ------------------------------

int main(int argc, char * argv[])
{
    // Initialize ROS2 Client Library for C++
    rclcpp::init(argc, argv);

    // Spin node by using shared pointer
    rclcpp::spin(std::make_shared<TurtleListener>());
    
    // When spin ends, shutdown and close
    rclcpp::shutdown();
    return 0;
}